File: python/src/assesslite/graph.py
"""Causal-graph check (AssessLite core spec v0.2, spec/graph/graph-check.md).

Declares a DAG, derives its implied conditional independencies (ordered local
Markov), and tests each against the data by partial correlation. Self-contained
numpy/pandas; no external graph package. Kept identical to the R engine.
"""
from __future__ import annotations

import math

import numpy as np
import pandas as pd

_COLS = ["label", "estimate", "se", "ci_low", "ci_high", "n"]


def parse_graph(edges) -> dict:
    """Parse edges like 'age -> adherence' into nodes, a parents map, and a topological order."""
    if not edges:
        raise ValueError("declare_graph needs at least one edge like 'a -> b'")
    frm, to = [], []
    for e in edges:
        parts = [p.strip() for p in str(e).split("->")]
        if len(parts) != 2 or not all(parts):
            raise ValueError(f"edge '{e}' is not of the form 'a -> b'")
        frm.append(parts[0]); to.append(parts[1])
    nodes = list(dict.fromkeys(frm + to))
    parents = {v: list(dict.fromkeys(f for f, t in zip(frm, to) if t == v)) for v in nodes}
    placed, remaining = [], list(nodes)
    while remaining:
        ready = [v for v in remaining if all(p in placed for p in parents[v])]
        if not ready:
            raise ValueError("the declared graph is not acyclic")
        placed += ready
        remaining = [v for v in remaining if v not in ready]
    return {"edges": list(edges), "nodes": nodes, "parents": parents, "order": placed}


def coerce_numeric(x: pd.Series):
    """Coerce to a single numeric column, or None if a multi-level categorical."""
    if pd.api.types.is_bool_dtype(x):
        return x.astype(float).to_numpy()
    if pd.api.types.is_numeric_dtype(x) and not isinstance(x.dtype, pd.CategoricalDtype):
        return x.astype(float).to_numpy()
    u = sorted(pd.unique(x.dropna().astype(str)))
    if len(u) == 2:
        return x.astype(str).map({u[0]: 0.0, u[1]: 1.0}).to_numpy()
    return None


def _design(Zdf: pd.DataFrame) -> np.ndarray:
    blocks = [np.ones((len(Zdf), 1))]
    for c in Zdf.columns:
        s = Zdf[c]
        if isinstance(s.dtype, pd.CategoricalDtype) or not pd.api.types.is_numeric_dtype(s):
            d = pd.get_dummies(s, prefix=c, drop_first=True)
            if d.shape[1] > 0:
                blocks.append(d.to_numpy(dtype=float))
        else:
            blocks.append(s.to_numpy(dtype=float).reshape(-1, 1))
    return np.hstack(blocks)


def partial_cor(v: np.ndarray, w: np.ndarray, Zdf):
    if Zdf is None or Zdf.shape[1] == 0:
        rv = v - v.mean(); rw = w - w.mean(); k = 0
    else:
        X = _design(Zdf)
        bv, _, rank, _ = np.linalg.lstsq(X, v, rcond=None)
        bw, _, _, _ = np.linalg.lstsq(X, w, rcond=None)
        rv = v - X @ bv; rw = w - X @ bw
        k = int(rank) - 1
    n = len(v)
    denom = rv.std() * rw.std()
    r = float(np.corrcoef(rv, rw)[0, 1]) if denom > 0 else 0.0
    return r, n, k


def test_graph_check(assessment, alpha: float = 0.05, effect_floor: float = 0.1) -> dict:
    g = assessment.graph
    if g is None:
        raise ValueError("graph_check needs a declared graph; call declare_graph() first")
    d = assessment.data
    order, parents = g["order"], g["parents"]
    imps = []

    for i, V in enumerate(order):
        preds = order[:i]
        par = parents[V]
        for W in [p for p in preds if p not in par]:
            cond = list(par)
            claim = "{} _||_ {} | {{{}}}".format(V, W, ", ".join(cond))
            vv = coerce_numeric(d[V]) if V in d.columns else None
            ww = coerce_numeric(d[W]) if W in d.columns else None
            if vv is None or ww is None:
                imps.append({"claim": claim, "conditioning": cond, "partial_r": None,
                             "p_value": None, "n": None, "status": "not_testable"})
                continue
            # complete cases across v, w, and conditioning columns
            frame = pd.DataFrame({"__v": vv, "__w": ww})
            Z = d[cond].reset_index(drop=True) if cond else None
            mask = frame.notna().all(axis=1)
            if Z is not None:
                mask = mask & Z.notna().all(axis=1)
            vvc = frame["__v"][mask].to_numpy()
            wwc = frame["__w"][mask].to_numpy()
            Zc = Z[mask].reset_index(drop=True) if Z is not None else None
            r, n, k = partial_cor(vvc, wwc, Zc)
            dfree = n - k - 3
            if not math.isfinite(r) or dfree < 1:
                status, p = "not_resolvable", None
            else:
                rc = max(min(r, 1 - 1e-12), -1 + 1e-12)
                z = math.atanh(rc) * math.sqrt(dfree)
                p = math.erfc(abs(z) / math.sqrt(2))
                mds = math.tanh(1.96 / math.sqrt(dfree))
                status = ("violated" if (p < alpha and abs(r) >= effect_floor)
                          else "not_resolvable" if mds > effect_floor else "consistent")
            imps.append({"claim": claim, "conditioning": cond, "partial_r": round(r, 4),
                         "p_value": None if p is None else round(p, 4), "n": int(n), "status": status})

    statuses = [im["status"] for im in imps]
    n_testable = sum(s != "not_testable" for s in statuses)
    if any(s == "violated" for s in statuses):
        verdict = "unstable"
    elif n_testable == 0 or any(s == "not_resolvable" for s in statuses):
        verdict = "not_resolvable"
    else:
        verdict = "stable"

    n_viol = sum(s == "violated" for s in statuses)
    if verdict == "unstable":
        first = next(im["claim"] for im in imps if im["status"] == "violated")
        reading = (f"the data contradict {n_viol} of the graph's {n_testable} implied independencies "
                   f"(e.g. {first}), so the declared causal graph does not hold up (partial-correlation "
                   f"test, a linear approximation to conditional independence)")
    elif verdict == "not_resolvable":
        reading = ("none of the graph's implied independencies were testable at this n (endpoints were "
                   "multi-level categoricals); the graph is not resolvable against this data"
                   if n_testable == 0 else
                   f"no implied independence was contradicted, but at least one of the {n_testable} "
                   f"testable implications was underpowered; the graph is not fully resolvable at this n")
    else:
        reading = (f"every one of the {n_testable} testable implied independencies is consistent with the "
                   f"data (partial-correlation test); the data do not contradict the declared graph. This "
                   f"does not establish the graph — Markov-equivalent graphs share these implications")

    return {"test": "graph_check", "invariance": "causal_graph", "verdict": verdict,
            "metrics": None, "implications": imps,
            "variants": pd.DataFrame(columns=_COLS), "n_failed": 0, "reading": reading}

File: R/graph.R
# Causal-graph check (AssessLite core spec v0.2, spec/graph/graph-check.md).
# Declares a DAG, derives its implied conditional independencies (ordered local
# Markov), and tests each against the data by partial correlation. Self-contained
# base R; no external graph package.

# Parse edges like "age -> adherence" into nodes, a parents map, and a topological
# order. Errors if the declared graph is cyclic.
parse_graph <- function(edges) {
  if (length(edges) == 0) stop("declare_graph() needs at least one edge like 'a -> b'")
  from <- to <- character(0)
  for (e in edges) {
    parts <- trimws(strsplit(e, "->", fixed = TRUE)[[1]])
    if (length(parts) != 2 || any(!nzchar(parts)))
      stop("edge '", e, "' is not of the form 'a -> b'")
    from <- c(from, parts[1]); to <- c(to, parts[2])
  }
  nodes <- unique(c(from, to))
  parents <- stats::setNames(lapply(nodes, function(v) unique(from[to == v])), nodes)
  # Kahn-style topological order
  placed <- character(0); remaining <- nodes
  while (length(remaining) > 0) {
    ready <- remaining[vapply(remaining, function(v) all(parents[[v]] %in% placed), logical(1))]
    if (length(ready) == 0) stop("the declared graph is not acyclic")
    placed <- c(placed, ready); remaining <- setdiff(remaining, ready)
  }
  list(edges = edges, nodes = nodes, parents = parents, order = placed)
}

#' Declare a causal DAG for the graph_check attack
#'
#' @param audit a structural_audit object.
#' @param edges character vector of directed edges, e.g. c("age -> adherence",
#'   "stage -> survival"). Nodes are the union of everything named.
declare_graph <- function(audit, edges) {
  stopifnot(inherits(audit, "structural_audit"))
  g <- parse_graph(edges)
  missing <- setdiff(g$nodes, names(audit$data))
  if (length(missing) > 0)
    warning("graph nodes not found in the data (implications involving them will be skipped): ",
            paste(missing, collapse = ", "))
  audit$graph <- g
  audit
}

# Coerce a variable to a single numeric column, or NULL if it is a multi-level
# categorical (not testable as an endpoint of an independence claim).
coerce_numeric <- function(x) {
  if (is.logical(x)) return(as.numeric(x))
  if (is.numeric(x)) return(as.numeric(x))
  u <- sort(unique(as.character(x[!is.na(x)])))
  if (length(u) == 2) return(as.numeric(factor(as.character(x), levels = u)) - 1)
  NULL
}

# Partial correlation of v and w given conditioning frame Z (may be NULL/empty).
partial_cor <- function(v, w, Z) {
  if (is.null(Z) || ncol(Z) == 0) {
    ok <- stats::complete.cases(v, w)
    v <- v[ok]; w <- w[ok]
    rv <- v - mean(v); rw <- w - mean(w); k <- 0L
  } else {
    ok <- stats::complete.cases(v, w, Z)
    v <- v[ok]; w <- w[ok]; Z <- Z[ok, , drop = FALSE]
    for (j in seq_along(Z)) if (is.character(Z[[j]])) Z[[j]] <- factor(Z[[j]])
    mm <- stats::model.matrix(~ ., data = Z)
    fv <- stats::lm.fit(mm, v); fw <- stats::lm.fit(mm, w)
    rv <- fv$residuals; rw <- fw$residuals; k <- fv$rank - 1L
  }
  n <- length(v)
  r <- suppressWarnings(stats::cor(rv, rw))
  list(r = r, n = n, k = k)
}

# graph_check: attacks the causal_graph invariance by testing the DAG's implied
# conditional independencies against the data.
test_graph_check <- function(audit, alpha = 0.05, effect_floor = 0.1) {
  g <- audit$graph
  if (is.null(g)) stop("graph_check needs a declared graph; call declare_graph() first")
  d <- audit$data; ord <- g$order; parents <- g$parents
  imps <- list()

  for (i in seq_along(ord)) {
    V <- ord[i]
    preds <- ord[seq_len(i - 1)]
    par <- parents[[V]]
    for (W in setdiff(preds, par)) {
      cond <- par
      claim <- sprintf("%s _||_ %s | {%s}", V, W, paste(cond, collapse = ", "))
      vv <- if (V %in% names(d)) coerce_numeric(d[[V]]) else NULL
      ww <- if (W %in% names(d)) coerce_numeric(d[[W]]) else NULL
      if (is.null(vv) || is.null(ww)) {
        imps[[length(imps) + 1]] <- list(claim = claim, conditioning = cond,
          partial_r = NA_real_, p_value = NA_real_, n = NA_integer_, status = "not_testable")
        next
      }
      Z <- if (length(cond) > 0) d[cond] else NULL
      pc <- partial_cor(vv, ww, Z)
      dfree <- pc$n - pc$k - 3
      if (!is.finite(pc$r) || dfree < 1) {
        status <- "not_resolvable"; p <- NA_real_
      } else {
        z <- atanh(pmax(pmin(pc$r, 1 - 1e-12), -1 + 1e-12)) * sqrt(dfree)
        p <- 2 * stats::pnorm(-abs(z))
        mds <- tanh(1.96 / sqrt(dfree))
        status <- if (p < alpha && abs(pc$r) >= effect_floor) "violated"
                  else if (mds > effect_floor) "not_resolvable" else "consistent"
      }
      imps[[length(imps) + 1]] <- list(claim = claim, conditioning = cond,
        partial_r = round(pc$r, 4), p_value = if (is.na(p)) NA_real_ else round(p, 4),
        n = pc$n, status = status)
    }
  }

  statuses <- vapply(imps, function(x) x$status, character(1))
  n_testable <- sum(statuses != "not_testable")
  verdict <- if (any(statuses == "violated")) "unstable"
             else if (n_testable == 0 || any(statuses == "not_resolvable")) "not_resolvable"
             else "stable"
  n_viol <- sum(statuses == "violated")
  reading <- switch(verdict,
    unstable = sprintf(paste0("the data contradict %d of the graph's %d implied independencies ",
      "(e.g. %s), so the declared causal graph does not hold up (partial-correlation test, a ",
      "linear approximation to conditional independence)"),
      n_viol, n_testable,
      imps[[which(statuses == "violated")[1]]]$claim),
    not_resolvable = if (n_testable == 0)
      "none of the graph's implied independencies were testable at this n (endpoints were multi-level categoricals); the graph is not resolvable against this data"
      else sprintf(paste0("no implied independence was contradicted, but at least one of the %d ",
        "testable implications was underpowered; the graph is not fully resolvable at this n"), n_testable),
    stable = sprintf(paste0("every one of the %d testable implied independencies is consistent with ",
      "the data (partial-correlation test); the data do not contradict the declared graph. This does ",
      "not establish the graph — Markov-equivalent graphs share these implications"), n_testable))

  list(test = "graph_check", invariance = "causal_graph", verdict = verdict, metrics = NULL,
       implications = imps,
       variants = data.frame(label = character(), estimate = numeric(), se = numeric(),
                             ci_low = numeric(), ci_high = numeric(), n = integer(),
                             stringsAsFactors = FALSE),
       n_failed = 0L, reading = reading)
}
